[IV-23-24] Objetivo 4

MarioGuisado opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
Sí
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
Sí, la validación previa y las funciones que sabemos que se utilizarán en la lógica final.
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
Sí, los propios tests.
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
Se evaluarán las clases EntrenadorDisponible y AtletaDisponible.
* [ ] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
Hay que leer hombre.
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
Se han seguido.
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

MarioGuisado committed:
Herramientas definidas para #37

MarioGuisado committed:
Herramientas seleccionadas para #37

MarioGuisado committed:
Modificado enumerado

MarioGuisado committed:
Comprobaciones en el constructor y métodos mínimos necesarios para validar #39

MarioGuisado committed:
Archivos de configuracion modificados

Archivos modificados para usar require en lugar de import siguiendo las buenas prácticas de node y su implementación original

MarioGuisado committed:
Imports sustituidos y funciones mejoradas para #39

MarioGuisado committed:
Tests añadidos para validar #39

MarioGuisado committed:
README.md e iv.yaml modificados

MarioGuisado committed:
tests.md modificado

Morad02 commented on docs/tests.md:
```diff
@@ -0,0 +1,33 @@
+-Jest: el framework de Facebook, aunque fue diseñado principalmente para aplicaciones basadas en React, también es posible utilizarlo con Node. Además cuenta con su propia biblioteca de aserciones.
+
+Todos los candidatos considerados cuentan con CLI.
+
```
Las características que describes de estos candidatos son muy generales y no creo que sirvan mucho para eligir las herramientas más razonables. Si vas a elegir Chai como biblioteca de aserciones, creo que deberías hablar un poco de él.

MarioGuisado replied:
Resuelto en https://github.com/MarioGuisado/TrainMe/pull/38/commits/9efd049154e7d5dd672f5f4b252b2def7b7c2c85

Morad02 commented on docs/tests.md:
```diff
@@ -0,0 +1,33 @@
+## Elección
+
+Aunque todos los candidatos serían válidos y suficientes para establecer nuestros tests, creo que la mejor opción sería el framework Mocha junto con la biblioteca de aserciones Chai. Esta combinación es una de las más populares en proyectos que emplean Node.js y TypeScript, dan una visión realista de como sería un proyecto "serio" y da la oportunidad de aprender a usar herramientas presentes en e ...
+
```
Si la popularidad de las herramientas va a ser un criterio de elección deberías decirlo en el apartado correspondiente. Además, creo que la explicación de la elección debería ir en base a los criterios que has puesto al principio

MarioGuisado replied:
Resuelto en https://github.com/MarioGuisado/TrainMe/pull/38/commits/9efd049154e7d5dd672f5f4b252b2def7b7c2c85

Morad02 commented on docs/tests.md:
```diff
@@ -0,0 +1,33 @@
+## Elección
+
+Aunque todos los candidatos serían válidos y suficientes para establecer nuestros tests, creo que la mejor opción sería el framework Mocha junto con la biblioteca de aserciones Chai. Esta combinación es una de las más populares en proyectos que emplean Node.js y TypeScript, dan una visión realista de como sería un proyecto "serio" y da la oportunidad de aprender a usar herramientas presentes en e ...
+
```
No se justifica el por qué del uso de la CLI de Mocha para ejecutar los tests

MarioGuisado replied:
Todos los candidatos incluyen CLI, el usado será el que incluya el framework seleccionado.

Morad02 commented on src/atleta.ts:
Siguiendo la filosofía de código limpio y SRP (single responsibility principle). Una función solo debe encargarse de hacer una cosa y tiene que caber en un solo pantallazo. Viendo el constructor, vemos que no cumple con estos requisitos.

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/2903dc724509b5e09cf7bf3af6c74ef93dfc6ca7

Morad02 commented on src/entrenador.ts:
```diff
@@ -1,30 +1,59 @@
         this._categoriasPeso = categoriasPeso;
-
     }
 }
```
Aquí pasa lo mismo,  no sigue las indicaciones del SRP y código limpio

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/2903dc724509b5e09cf7bf3af6c74ef93dfc6ca7

Morad02 reviewed (changes requested):
En líneas generales, creo que los tests están bien.  Corrigiendo los commits y lo comentado creo que se estaría más cerca del aprobado.

MarioGuisado committed:
Criterios redefinidos para #37

MarioGuisado committed:
Constructor corregido para #39

Constructor corregido siguiendo SRP

JJ commented on iv.yaml:
```diff
@@ -5,3 +5,4 @@ entidad:
 automatizar:
   fichero: package.json
   orden: npm run 
+test: test/atletadisponible_expect.ts test/entrenadordisponible_expect.ts
```
Sólo debes poner un fichero de test.

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/9a91a20c6e144ff30e2e7a37919c8eeeb648e152

JJ reviewed (changes requested):
`npm run test` no funciona

```txt

> iv@1.0.0 test
> mocha test/*


Error: Cannot find module '../dist/entrenadordisponible.js'
Require stack:
- /home/jmerelo/Asignaturas/infraestructura-virtual/TrainMe/test/atletadisponible_expect.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Module._load (node:internal/modules/cjs/loader:901:27)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:119:18)
    at Object.<anonymous> (/home/jmerelo/Asignaturas/infraestructura-virtual/TrainMe/test/atletadisponible_expect.js:2:30)
    at Module._compile (node:internal/modules/cjs/loader:1233:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1287:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:165:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:192:25)
    at async DefaultModuleLoader.import (node:internal/modules/esm/loader:228:24)
    at async importModuleDynamicallyWrapper (node:internal/vm/module:428:15)
    at async formattedImport (/home/jmerelo/Asignaturas/infraestructura-virtual/TrainMe/node_modules/mocha/lib/nodejs/esm-utils.js:9:14)
    at async exports.requireOrImport (/home/jmerelo/Asignaturas/infraestructura-virtual/TrainMe/node_modules/mocha/lib/nodejs/esm-utils.js:42:28)
    at async exports.loadFilesAsync (/home/jmerelo/Asignaturas/infraestructura-virtual/TrainMe/node_modules/mocha/lib/nodejs/esm-utils.js:100:20)
    at async singleRun (/home/jmerelo/Asignaturas/infraestructura-virtual/TrainMe/node_modules/mocha/lib/cli/run-helpers.js:125:3)
    at async exports.handler (/home/jmerelo/Asignaturas/infraestructura-virtual/TrainMe/node_modules/mocha/lib/cli/run.js:370:5)
```

Tendrás que asegurarte de que mocha funcione bien con typescript; del tirón no tiene por qué hacerlo.

MarioGuisado committed:
Corregido iv.yaml #37

Corregido iv.yaml para hacer referencia a un único fichero de test

MarioGuisado committed:
Archivos de configuración modificados para #37

MarioGuisado committed:
Atleta.ts modificado para #39

MarioGuisado committed:
Test creados para #37

MarioGuisado committed:
Test modificado para #37

MarioGuisado committed:
refactor: cambiados require por import (#41)

MarioGuisado committed:
Delete test/atletadisponible_expect.js

MarioGuisado committed:
Delete test/entrenadordisponible_expect.js

JJ commented on src/atleta.ts:
```diff
@@ -27,4 +36,43 @@ export class Atleta{
+        }
+    }
+
+    private validarModalidadEntreno(modalidadEntreno: Map<number, string>): void {
```
¿Seguro que no puedes hacer un poco de DRY aquí? ¿No se trata de validar cosas en un rango al fin y al cabo?

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/59a30851bed890cd5317371184f8234495b92983

JJ commented on src/entrenadordisponible.ts:
```diff
@@ -15,11 +16,48 @@ export class EntrenadorDisponible extends Entrenador {
         this._id = id;
+        this._atletasSugeridos = [];
+        this._atletasElegidos = [];
+        this._disponibilidad = true;
```
Estás usando como variables de instancia cosas que posiblemente deberían ser sólo llamadas a función. Igual, estás convirtiendo en una entidad un solo objeto valor, y lo habitual es que una entidad gestione agregados de objetos valor.

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/e8705558f25a233e2e52f0c2b1d51e79c1c475f8 Ahora que lo comentas creo que es mejor manejar estos agregados en la futura implementación final de la lógica de negocio. Por eso he eliminado los métodos relacionados con estas variables (que como dices solo debería ser llamadas a función) y he modificado los tests consecuentemente en https://github.com/MarioGuisado/TrainMe/pull/38/commits/1208e979cbc4133b4afc3a839b8232f88c9a86ce de forma que solo compruebo lo esencial antes de empezar a implementar dicha lógica.

JJ reviewed (changes requested):
El problema es que creo que has complicado el código un poco más de la cuenta, con lo que es complicado saber realmente si has seguido las HUs. Deberías tratar de simplificarlo; si tienes los tests y funcionan bien, deberían seguir funcionando.

MarioGuisado committed:
Aplicado DRY, closes #42

MarioGuisado committed:
Variables de instancia redefinidas, closes #43

MarioGuisado committed:
Test redefinidos para #39, ateniendose a los cambios realizados para resolver #43

Solo se comprueban las construcciones de los objetos y los métodos relacionados con sus variables de instancia

JJ commented on iv.yaml:
```diff
@@ -5,3 +5,4 @@ entidad:
 automatizar:
   fichero: package.json
   orden: npm run 
+test: test/atletadisponible_expect.ts
```
¿Añadirle "expect" es la buena práctica en este sentido?

MarioGuisado replied:
Modificado en https://github.com/MarioGuisado/TrainMe/pull/38/commits/ddae5ed6f9af583fca3da11afe4277ef26c82e20

JJ commented on package.json:
```diff
@@ -8,15 +8,21 @@
-    "check": "tsc --noEmit src/*",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "check": "tsc --project . --noEmit",
+    "test": "env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' mocha -r ts-node/register 'test/*.ts'"
```
¿Seguro que no hay otra forma de configurarlo que no sea con una variable temporal de entorno? ¿No lo puedes poner en tsconfig.json?

MarioGuisado replied:
Resuelto en https://github.com/MarioGuisado/TrainMe/pull/38/commits/3c38aa9ff7e925c4f0880efe3c60e0f521eb4228

JJ commented on package.json:
```diff
@@ -8,15 +8,21 @@
+    "@types/chai": "^4.3.10",
+    "@types/mocha": "^10.0.4",
+    "@types/node": "^20.9.0",
     "@typescript-eslint/eslint-plugin": "^6.9.1",
```
Si no lo usas, ¿por qué no lo has desinstalado?

MarioGuisado replied:
Resuelto en https://github.com/MarioGuisado/TrainMe/pull/38/commits/3c38aa9ff7e925c4f0880efe3c60e0f521eb4228

JJ commented on src/atletadisponible.ts:
```diff
@@ -1,25 +1,48 @@
+        return this._disponibilidad;
+    }
+
+    asignarDisponibilidad(disponibilidad: boolean): void {
```
Para empezar, esta clase se llama `AtletaDisponible`. Si puede o no estarlo, el nombre no sería correcto. Pero el problema mayor es lo ya señalado varias veces en el objetivo 2 (y que no sé por qué se me pasó aquí en concreto): los objetos valor deben ser inmutables, y si hay algún cambio de estado no pueden manejarlo ellos, sino una entidad de orden superior. Mira por favor [este error frecuente](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md):

> Los objetos valor son inmutables; no pueden incluir ninguna variable de instancia que indique que han cambiado de estado, por ejemplo disponible o completada. El hacerlo así indica un error en la modelización del dominio: son las entidades las que tienen que gestionar esos cambios de estado, usando las estructuras correspondientes que liguen el objeto valor (inmutable) con su cambio de estado.

Y mis disculpas por no haberlo visto antes.

MarioGuisado replied:
Clase eliminada en https://github.com/MarioGuisado/TrainMe/pull/38/commits/52e53129ff8e38a684c9d16649026938917b94cf

JJ commented on src/atletadisponible.ts:
```diff
@@ -1,25 +1,48 @@
+    }
+
+    asignarDisponibilidad(disponibilidad: boolean): void {
+        if (typeof disponibilidad !== 'boolean') {
```
TypeScript comprueba los tipos para que tú no tengas que hacerlo.

MarioGuisado replied:
Clase eliminada en https://github.com/MarioGuisado/TrainMe/pull/38/commits/52e53129ff8e38a684c9d16649026938917b94cf

JJ commented on src/entrenadordisponible.ts:
```diff
@@ -15,11 +13,27 @@ export class EntrenadorDisponible extends Entrenador {
+        return this._disponibilidad;
+    }
+
+    asignarDisponibilidad(disponibilidad: boolean): void {
```
Exactamente igual que antes. ¿Te das cuenta que estás generando una clase nueva y código nuevo para algo que se podría gestionar directamente desde una entidad?

MarioGuisado replied:
Clase eliminada en https://github.com/MarioGuisado/TrainMe/pull/38/commits/fcd23769f0b8624964c28f1b60a99880b41fc3d5

JJ commented on src/entrenadordisponible.ts:
```diff
@@ -15,11 +13,27 @@ export class EntrenadorDisponible extends Entrenador {
+    }
+
+    asignarDisponibilidad(disponibilidad: boolean): void {
+        if (typeof disponibilidad !== 'boolean') {
```
... y además repitiendo código.

MarioGuisado replied:
Clase eliminada en https://github.com/MarioGuisado/TrainMe/pull/38/commits/fcd23769f0b8624964c28f1b60a99880b41fc3d5

JJ reviewed (changes requested):
El problema principal, aparte de lo planteado, es que realmente no has implementado la lógica de negocio. Deberías haber creado un issue para hacerlo, y tratar de resolver el problema aquí. 
De hecho, los tests no comprueban nada relacionado con historias de usuario, sólo asignan valores y ven si los valores están bien asignados (que lo van a estar). Ni siquiera comprueban posibles errores o excepciones.

MarioGuisado committed:
Archivo de cofiguración redefinido, closes #45

MarioGuisado committed:
Clases redefinidas, closes #47

MarioGuisado committed:
Clase emparejador, closes #48

MarioGuisado committed:
Test para el emparejador, closes #39

MarioGuisado committed:
iv.yaml modificado, closes #49

MarioGuisado committed:
Delete src/atletadisponible.ts

MarioGuisado committed:
Delete src/entrenadordisponible.ts

MarioGuisado committed:
Delete test/entrenadordisponible_expect.ts

MarioGuisado committed:
Delete test/atletadisponible_expect.ts

JJ commented on src/atleta.ts:
```diff
@@ -1,30 +1,50 @@
-
+    }
+ 
+    getid(): number {
```
Usa CamelCase aquí.

MarioGuisado replied:
Aunque la clase ha sido eliminada, he incluido CamelCase en https://github.com/MarioGuisado/TrainMe/pull/38/commits/7e89f10b4d513a6de44272195ead7ec95a001c69

JJ commented on src/atleta.ts:
```diff
@@ -1,30 +1,50 @@
+    getid(): number {
+        return this._id;
+    }
+    getfechaNacimiento(): Date {
```
En JS se pueden declarar los getters con `get <funcion>`. Si no lo haces, la vocal después de `get` tiene que ir en mayúsculas.

MarioGuisado replied:
La clase ha sido eliminada en https://github.com/MarioGuisado/TrainMe/pull/38/commits/c46b36d275cd266f2952ce4dded300a04bddeb99

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,109 @@
+    vector.push(usuario.getid());
+
+
+    vector.push(usuario.getnivelRendimiento().values().next().value);
```
No entiendo muy bien. Si estás usando un `Map`, ¿por qué no estás accediendo por clave a cada valor? Si vas a acceder sólo a los valores, ¿por qué no usas un array?

MarioGuisado replied:
La verdad que el map me ha complicado de más las cosas, sobre todo viendo que luego solo iba a asignar un único valor a cada atributo. Los tipos estan corregidos en https://github.com/MarioGuisado/TrainMe/pull/38/commits/8efa73ea0cb658feaaf61c51afb9924e7a5bc8f6

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,109 @@
+
+
+    vector.push(usuario.getnivelRendimiento().values().next().value);
+    vector.push(usuario.getnivelCompromiso().values().next().value);
```
Tampoco acabo de entender este `next`. ¿Estás usando un iterador que guarda el estado del cursos de una llamada a `procesarUsuario` a otra?

MarioGuisado replied:
Clase corregida en https://github.com/MarioGuisado/TrainMe/pull/38/commits/2359ce8b9e0cc801ee4ab7f41ceac49ea3ce26ab

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,109 @@
+
+    vector.push(usuario.getnivelRendimiento().values().next().value);
+    vector.push(usuario.getnivelCompromiso().values().next().value);
+    vector.push(usuario.getmodalidadEntreno().values().next().value);
```
Es que no entiendo muy bien. ¿Por qué hay un `Map` si sólo usas un valor y ni siquiera la clave?

MarioGuisado replied:
Tipos corregidos en https://github.com/MarioGuisado/TrainMe/pull/38/commits/8efa73ea0cb658feaaf61c51afb9924e7a5bc8f6

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,109 @@
+    let minDistance = Infinity;
+    let distance;
+    let matriz;
+    let id_elegido = -1;
```
Esto *nunca* es buena costumbre. Usar un flag para indicar si se ha encontrado algo o no es un antipatrón. 
Puedes usar en TypeScript `filter` para filtrar, en vez de un bucle, pero nunca esto.

MarioGuisado replied:
Clase corregida en https://github.com/MarioGuisado/TrainMe/pull/38/commits/2359ce8b9e0cc801ee4ab7f41ceac49ea3ce26ab

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,109 @@
+      }
+    });
+    
+    return id_elegido;
```
¿Y si es -1? Estás devolviendo un valor inválido.
Además, Javascript tiene la función `sort`

MarioGuisado replied:
Clase corregida en https://github.com/MarioGuisado/TrainMe/pull/38/commits/2359ce8b9e0cc801ee4ab7f41ceac49ea3ce26ab

JJ commented on src/entrenador.ts:
```diff
@@ -1,30 +1,59 @@
-    private _categoriasPeso: Map<string, [number, number]>;
+    private _nivelRendimiento: Map<string, number>;
+    private _nivelCompromiso: Map<string, number>;
+    private _modalidadEntreno: Map<string, number>;
```
Si estas variables de instancia son las mismas para Entrenador y Atleta, ¿por qué no usas la misma clase? Para eso está la herencia. Aunque en este caso ni siquiera tengo muy claro que realmente haya alguna diferencia, salvo un campo que siga si son entrenadores o jugadores o ni siquiera eso, simplemente que estén en diferentes arrays en el emparejador.

MarioGuisado replied:
Clase eliminada en https://github.com/MarioGuisado/TrainMe/pull/38/commits/c46b36d275cd266f2952ce4dded300a04bddeb99

JJ commented on src/types.ts:
```diff
@@ -1,33 +1,38 @@
-    [0, 'Whatsapp'],
-    [1, 'Telegram'],
-    [2, 'Email'],
+export const TipoContacto = new Map<string, number>([
```
¿De veras que vas a asignarle valor numérico al TipoContacto? ¿No deberías simplemente filtrar?

MarioGuisado replied:
Los tipos están corregidos en https://github.com/MarioGuisado/TrainMe/pull/38/commits/8efa73ea0cb658feaaf61c51afb9924e7a5bc8f6

JJ reviewed (changes requested):
Me temo que hay una serie de problemas con el código y deberías de considerar una serie de alternativas, partiendo de la reformulación de los issues (si es necesario). En algunos casos, tendrás también que aplicar buenas prácticas.

MarioGuisado committed:
Clases redefinidas, closes #47

MarioGuisado committed:
Merge branch 'Objetivo-4' of github.com:MarioGuisado/StudyTogether into Objetivo-4

MarioGuisado committed:
Corregida clase emparejador, closes #48

MarioGuisado committed:
Corregido types.ts, closes #50

MarioGuisado committed:
Corregido tests, closes #39

MarioGuisado committed:
Incluído CamelCase, closes #51

MarioGuisado committed:
Comentarios eliminado en types.ts para #50

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,102 @@
+    const vector: number[] = [];
+
+    vector.push(usuario.GetId());
+    vector.push(usuario.GetNivelRendimiento());
```
Sigues repitiendo código. Y metiendo en el vector de la misma forma ¿el ID?

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/issues/48 El emparejador guarda los datos de cada usuario y reserva la primera columna para el ID, de forma que se identifica fácilmente al usuario. No se utiliza para calcular la distancia.

JJ replied:
Tampoco es una buena práctica poner cosas diferentes en un array. Si no lo usas para calcular la distancia, usa otro método para identificar de quién se está hablando (como un hash)

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/515b37dc19085fb15122db15b7c0663b58d9a758

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,102 @@
+    vector.push(usuario.GetNivelRendimiento());
+    vector.push(usuario.GetNivelCompromiso());
+    vector.push(usuario.GetModalidadEntreno());
+    vector.push(usuario.GetPreferenciasContacto());
```
No creo que la preferencia contacto puedas incluirla en la medición de distancia. ¿No será simplemente una forma de filtrar unos y otros?

MarioGuisado replied:
Todos los datos (menos el ID) se pueden utililizar para calcular distancia (un atleta puede elegir que un entrenador con un nivel de compromiso distinto al suyo le sea sugerido), o se pueden utilizar para filtrar (solo los entrenadores con igual nivel de compromiso le serán sugeridos), creo que es lo que más sentido tiene en lugar de separar en datos para calcular distancia, y datos para filtrar, o al menos es lo que en una situación "real" me parece más natural. Las opciones de filtrado se han definido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/d37d539eb3c2a02ff3ebdfe6a30bdbd42e6dd81d y se han ajustado las funciones del emparejador para usar los filtros en  https://github.com/MarioGuisado/TrainMe/pull/38/commits/4197b8e49b4bb956ec846e229a99f48efdbfa1cd

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,102 @@
+    vector.push(usuario.GetId());
+    vector.push(usuario.GetNivelRendimiento());
+    vector.push(usuario.GetNivelCompromiso());
+    vector.push(usuario.GetModalidadEntreno());
```
La modalidad ¿es una forma de filtrar o una cantidad de la cual se pueda aplicar una distancia?

MarioGuisado replied:
Ambos

JJ commented on src/usuario.ts:
```diff
@@ -0,0 +1,39 @@
+        id: number) 
+    {
+        this._id = id;
+        this._nivelRendimiento = nivelRendimiento;
```
Por favor, Mario, recuerda lo que hablamos en el objetivo 2. Había algunas variables que servían para filtrar, y otras para calcular distancias. Estś metiendo todo en el mismo saco, hasta el id, para calcular distancias. ¿De veras que se elige el entrenador de esa forma?

MarioGuisado replied:
Corregido en https://github.com/MarioGuisado/TrainMe/pull/38/commits/515b37dc19085fb15122db15b7c0663b58d9a758

JJ reviewed (changes requested):
Gracias por simplificar el código, pero me temo que no se trata tanto de simplificar como de usar las estructuras de datos más adecuadas a la lógica de negocio, tal como llevamos hablando desde el objetivo 2.

MarioGuisado committed:
Filtros definidos, closes #52

MarioGuisado committed:
emparejador corregido, closes #48

MarioGuisado committed:
test corregido, closes #39

MarioGuisado committed:
Filtros corregidos, closes #52

MarioGuisado committed:
Emparejador corregido, closes #48

MarioGuisado committed:
Test corregido, closes #39

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,108 @@
+    const vector: number[] = [usuario.GetNivelRendimiento(), usuario.GetNivelCompromiso(),
+    usuario.GetModalidadEntreno(), usuario.GetPreferenciasContacto()];
+
+    if(es_entrenador)
```
Si esto fuera la clave de un hash, no tendrías que hacer un if sino usarlo directamente para recuperar los valores

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,108 @@
+    let id_elegido;
+    let usuario: number[];
+
+    if (matriz_a_comparar == TipoMatriz.ENTRENADORES){
```
Igual, si usaras un hash indexado por el tipo podrías ahorrarte este if.

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,108 @@
+    }
+    
+    let indices_filtro: number[] = [];
+    filtros.forEach((filtro) => {
```
Si usaras un map en vez de un `forEach` te devolverían los elementos directamente en vez de los índices.

JJ commented on src/emparejador.ts:
```diff
@@ -0,0 +1,108 @@
+    return id_elegido;
+  }
+
+  GetUsuario(id_elegido: number, matriz_a_comparar: TipoMatriz): number[] | undefined {
```
en  CamelCase nuncas se pone la primera letra en mayúscula salvo qeu sea una clase.

JJ reviewed (changes requested):
Gracias por haberle dado una vuelta, creo que ha mejorado bastante. Creo que lo voy a dar por bueno, pero los dos errores son señalados son antipatrones importantes, así que convendría que los corrigieras antes de fusionar este PR o en todo caso antes del objetivo siguiente.

MarioGuisado committed:
Correcciones
